viewing_party/party.py

```diff
@@ -1,23 +1,193 @@
+            else:
+                genre_dict[movie['genre']] += 1
+        return max(genre_dict, key = genre_dict.get)
+
+    else:
+        return None
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+# we want to ensure uniqueness by grabbing/comparing both title and genre.
+# we want to have a clean comparison by not grabbing rating.
+
+def create_user_watched_set(user_data):
+
+    # I was on the fence about creating the 'watched' variable here.
+    # The function's pretty clear without it and we only have to use it once
+    # but might it be cleaner to use it since I assign this variable in almost all
+    # of my earlier functions? I found having the variable there helpful for the debugging
+    # watch list, for example, because it was always already set up. Thoughts appreciated.
+
+    user_watched = set()
+
+    for movie in user_data['watched']:
+        user_watched.add((movie['title'], movie['genre']))
+
+    return user_watched
+
+def create_friends_watched_set(user_data):
+    friends_watched = set()
+
+    for friend in user_data['friends']:
+        for movie in friend['watched']:
+            friends_watched.add((movie['title'], movie['genre']))
+    
+    return friends_watched
+
+def get_unique_watched(user_data):
+    watched = user_data['watched']
+    unique_movies = []
+
+    user_unique_watched = (create_user_watched_set(user_data)
+        - create_friends_watched_set(user_data))
+
+    for movie in watched:
+        for entry in user_unique_watched:
+            if entry == (movie['title'], movie['genre']):
+                unique_movies.append(movie)
+
+    return unique_movies
+
+def get_friends_unique_watched(user_data):
+    friends = user_data['friends']
+    friends_unique_movies = []
+
+    friends_unique_watched = (create_friends_watched_set(user_data) - 
+        create_user_watched_set(user_data))
+    
+    for friend in friends:
+        for movie in friend['watched']:
+            for entry in friends_unique_watched:
+                if entry == (movie['title'], movie['genre']):
+                    if movie in friends_unique_movies:
+                        continue
+                    else:
+                        friends_unique_movies.append(movie)
+
+    return friends_unique_movies
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    
+    # making sure that this is robust to subscriptions not existing because
+    # it's not in the Arrange section of the test where Sonya has no friends.
+    
+    subscriptions = user_data.get('subscriptions', [])
+    available_recs = []
+
+    friends_unique = get_friends_unique_watched(user_data)
+
+    for movie in friends_unique:
+        if movie['host'] in subscriptions:
+            available_recs.append(movie)

```

Comment:
We haven't talked about list comprehensions in python, but there's a special syntax that lets us effectively append a collection of values very concisely.

The following would initialize `available_recs` to be a list containing the movies from the friends' unique movies list that are in the user's subscription list.

```py
    available_recs = [movie for movie in friends_unique if movie['host'] in subscriptions]
```

The general pattern lets us take some code like
```py
new_list = []
for item in old_list:
    if should_process(item):
        new_list.append(process(item))
```

and write it like:
```py
new_list = [process(item) for item in old_list if should_process(item)]
```